### ISSUE TITLE
Refactor API Guide

### ISSUE THREAD

**Issue by catlook** [rackerlabs/docs-cloud-networks#110]
Instructions/resources here:
https://github.com/rackerlabs/docs-common/blob/master/refactor-api-guide.md
and
rackerlabs/docs-common#68
and
https://github.com/rackerlabs/docs-common/tree/master/api-guide-template
# Refactor API Guide checklist

Use the following checklist to guide and track the refactoring work for
each Developer Guide. If you have questions, open a [docs-common issue](https://github.com/rackerlabs/docs-common/issues/new).

When you build the refactored Sphinx project, use the `make clean html` command to build chunked content.  Before you begin refactoring, review the [restructured text fixes](#rst-coding-syntax-fixes).

**Important:** If the project contains additional source files or content significantly
different than the template, add a link to the source in [Issue #62 in the docs-common repo](https://github.com/rackerlabs/docs-common/issues/62).
## RST coding syntax fixes

Refactoring the API guides gives us a chance to make our RST source files comply with coding best practices.  You can make the following changes when you are refactoring, or make the updates in a separate effort. 
- Wrap all lines to 79 characters. 
- Ensure that heading separator lines are the same length as the title

```
    ====
    Test
    ====
```
- Use the following symbols for heading separators:
  - H1: over and underline `=`
  - H2: `~`
  - H3: `-`
  - H4: `^`

In the same file, the header levels must be in sequential order. If they aren't, you get the following error when you run the Sphinx build: `SEVERE/4) Title level inconsistent`.

**Tip:**  If you are using the Atom editor, you can configure the soft wrap feature to reflow text 
to 79 characters with a keybinding:
- Set the line length:
  -  Click **Atom > Preferences > Settings**.
  -  Change the **Preferred Line Length** to 79. 
  -  Make sure that **Soft wrap at Preferred Line length** is selected.
  -  Do not select the **Soft wrap** selection.
- Reflow text that exceeds 79 characters.
  - Select the line or paragraph.
  - Click **alt+cmd+q** to reflow the text. If text is indented or has special   formatting,
    you have to adjust the text manually.

**Note:** Don't worry about the line length in code samples or tables for now.
## Configuration updates (`conf.py`)
- [x] Replace everything above the `## -- General configuration` line with the first 32 lines from 
  the [`conf.py` in the API Guide template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/conf.py).
- [x] Change [deconst builder configuration](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/conf.py#L68) to `deconst-serial` to build chunked html.
       `builder = 'deconst-serial'`
       `#builder = 'deconst-single`
- [x] Add the [spelling checker extension](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/conf.py#L50) after the extension modules configuration.
- [x] Ensure that `extlinks` library matches the one in  [`conf.py` template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/conf.py#L122).
- [x] Add [`sphinx-rtd-template`](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/conf.py#L172) to improve template for local builds.
- [x] Add [requirements.txt](https://github.com/rackerlabs/docs-common/blob/master/requirements.txt) to the root directory to support local builds
- [x] Update titles to remove "Developer Guide". Follow title pattern in [`conf.py` template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/conf.py).
- [x] Verify that configuration settings match those in the conf.py for the API Guide template.
- [x] Check conf.py syntax with [pep8](http://pep8online.com/checkresult). Fix errors.

**Tip:** To catch errors inline, [configure Atom for Python](http://www.marinamele.com/install-and-configure-atom-editor-for-python).

After you complete the `conf.py` updates, run the Sphinx build (`make clean html`) to test that the build still works: `make clean html`. Resolve any errors.

**Note:** If the Sphinx build doesn't run, follow the [build from source instructions](https://github.com/rackerlabs/docs-rackspace/blob/master/doc/tools/build-from-source.rst) to install the required dependencies.
## Overview updates

When you refactor the content in the `overview` source files, you relocate the content and delete the `overview` folder and files. 
- [x] Move the following files to the root directory of the Sphinx project: `additional-resources.rst`,`service-updates.rst`
- [x] Copy the _About API_ description from `overview/index.rst` to the beginning of the main `index.rst` file so that the API description is the first information in the API Guide.
- [x] Move any other files in the `overview` folder to the root directory, and either integrate the content into another topic, or add the file in the `index.rst` toctree where it makes sense.
- [x] Delete the `overview` folder.
- [x] Update the `index.rst` _toctree_ directive:
  - [x] Delete `overview/index.rst`.
  - [x] Add `additional-resources` and `service-updates` after the `release-notes` file 
            and before the copyright as shown in the [index.rst template](https://github.com/rackerlabs/docs-common/edit/master/api-guide-template/index.rst).

After you complete the `overview` updates, run the local Sphinx build. Resolve any errors. Then, run the following command to review the local build output to verify that the API description, _Additional resources_ and _Service updates_ content renders correctly:  `open _build/html/index.html`
## Getting started updates

For the Getting Started section, use the following instructions to refactor the common content and the product-specific Getting Started content.
### Update common gs-section
- [x] Copy the `common-gs` folder with the [API Guide template files].(https://github.com/rackerlabs/docs-common/tree/master/api-guide-template/common-gs)
- [x] View the diff of the new content against your existing content.
  
    If you have any content that's not included in the common content, determine whether it
   needs to be added to the template.  If not, then put the content in a different location so
    that the common content is the same across all API Guides. If necessary, open a 
   [docs-common issue](https://github.com/rackerlabs/docs-common/issues/new) issue
   to update the template or determine  how to handle the custom content. All `common-gs`
  content must be the same.

The changes made to the common-gs content include the following:
- [x] Adjust heading levels in `auth-using-curl.rst` to move everything up one level.
- [x] Update get-credentials to move heading up a level and remove Cloud Control Panel login info.
- [x] How-to-use-curl minor edits and add Windows cURL info.
- [x] Wrap all lines at 79 characters and remove trailing spaces.
- [x]  Miscellaneous editorial improvements.
### Update Getting Started section
- [x] Add the [get-credentials-include.rst](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/getting-started/get-credentials-include.rst) file to the getting-started folder. This  file embeds the common `get-credentials.rst` topic at the H2 level.
- [x] Move `concepts.rst` file from the root directory to the `getting-started` folder. 
- [x] Update `authentication.rst` to append Identity endpoint information at end of last paragraph. See [authentication template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/getting-started/authenticate.rst).
- [x] Update `send-request-ovw.rst` to change intro sentence to refer to _section_ and
  remove link definition for Cloud Control panel. See [Send API request template](https://github.com/rackerlabs/docs-common/edit/master/api-guide-template/getting-started/send-request-ovw.rst).
- [x] Update `index.rst` ([See template.](https://github.com/rackerlabs/docs-common/edit/master/api-guide-template/getting-started/index.rst)) 
  - [x]  Copy the heading and opening paragraph from `getting-started.rst` to the
          top of `getting-started/index.rst`.
  - [x] Remove bold format from heading.
  - [x] Add the prerequisite information after the opening paragraph following the [getting-started topic template](https://github.com/rackerlabs/docs-common/edit/master/api-guide-template/getting-started/index.rst)
  - [x] Update `toctree` specification:
         \* Add `get-credentials-include` file at the top of the list. 
         \* Remove `prerequisites` topic.
         \* Add `concepts.rst` after `authentication.rst` topic.

After you complete the Getting Started updates, rebuild the Sphinx project. Resolve any errors. Then, review the local build output to verify that the content renders correcctly and that it matches the  [Getting started template](https://staging.developer.rackspace.com/staging.horse/build-71df1756f4/docs/api-doc-template/getting-started/).  
## Developer Guide updates

The Developer Guide refactoring moves the content from the Developer Guide introduction to the main `index.rst` file.
- [x] `index.rst` updates:
      - [x] Copy Developer Guide intro about audience and prerequisite knowledge to root `index.rst` intro before the paragraph with links to the sections of the API documentation. See [root index.rst template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/index.rst).
      - [x] Remove the link to the Developer Guide section.
      - [x] Update the _toctree_ specification to remove `developer-guide`
- [x] Delete the `developer-guide.rst` file from the root directory

After you complete the Developer Guide updates, rebuild the Sphinx project. Resolve any errors. Then, preview the local build output to verify that the Developer Guide section was removed.
## General API info updates

When you refactor the _General API section_, the main changes are removing the authentication
topic and adjusting the heading levels. You'll also need to change some common file names and content to match the common content in the API Guide template. Some of these changes are included in the instructions, but there are too many to document in detail. Be careful when you make these changes, some projects have custom information that differs from other projects. 

**Important:** If you have common files in your source that are not in the template directory or significantly deviate from the template source file, add a link to the Github source for that file to the [General API info source file issue](https://github.com/rackerlabs/docs-common/issues/62). This list will help us review and improve the General API Info content and determine whether it needs to be included in the API template.
- [x]  Change `index.rst` heading level to 

```
=======================
General API information
=======================
```
- [x] Update `general-api-info/index.rst`to append a reference to the Getting Started authentication section. (See [General API info index.rst  #template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/general-api-info/index.rst).)
- [x] Update the toctree directive to remove the filename for the authentication topic.
- [x] Delete the authentication source file from the `general-api-info` directory.
- [x] Change headings in sub-topics to H1  and H2 (~~~~~) per [template](https://github.com/rackerlabs/docs-common/edit/master/api-guide-template/general-api-info/api-behavior.rst).
- [x] Rename `service-access-endpoints.rst` to `service-access.rst` and match content to template version.
- [x] Compare _General API Information_ source files to the source in the API template directory, and update to match common content where possible. 

After you complete the Getting Started updates, rebuild the Sphinx project. Resolve any errors, Then, review the local build output to verify that the content renders correctly and that it matches the [General API information template](https://staging.developer.rackspace.com/staging.horse/build-71df1756f4/docs/api-doc-template/general-api-info/).
## API Reference updates

When you refactor the _API reference_ section, the main changes are relocating the api-reference heading to `api-reference/index.rst` topic and adjusting the heading levels in the methods section.
- [x] Rename `api-operations` folder to `api-reference`
- [x] Update the `exclude_pattern` configuration in `conf.py` with the new name.
- [x] Update `api-reference/index.rst` to follow [template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/api-reference/index.rst) 
- [x] Check  `api-reference.rst` in the root directory to make sure it doesn't have any additional
  information not included in the updated `api-reference/index.rst` file. If no extra content, delete the file.  If additional content is included and still needed, move it to the `index.rst` file.
- [x] Update root `index.rst` with new path to `api-reference/index.rst`.
- [x] Change api resource topic heading levels to h1 (====).
- [x] Change api method heading levels to h2 (~~~~~~~~~).
- [x] Change request and response heading levels to h3 (-------).

After you complete these updates, rebuild the Sphinx project. Resolve any errors. Then, review the local build output to verify that the content renders correctly and that it matches the API Reference  [API reference template](https://staging.developer.rackspace.com/staging.horse/build-71df1756f4/docs/api-doc-template/api-reference/).  
## Release notes updates

When you refactor the _API reference_ section, the main change is relocating the _Release notes_ heading and introduction to `release-notes/index.rst`, moving the release note topics to a sub-folder, and including all release notes topic in index.rst.  You also need to add the new `release-notes` folder to the `exclude_patterns` specification in the Sphinx configuration file, `conf.py`. 
- [x] Set up all release note content to render in a single topic.
  - [x]  Create a `releases` sub-folder in the `release-notes` directory.
  - [x]  Move individual release notes files into the `releases` folder.
  - [x]  Update `release-notes/index.rst` to  include all release files.
            Use the `.. include::` directive
- [x] `release-notes/index.rst updates:
      - [x] Change refid to`.._release-notes-collection:`.
      - [x]  Update`release-notes/index.rst`to  include all release files
              using the`.. include::`` directive. See [template](https://github.com/rackerlabs/docs-common/edit/master/api-guide-template/release-notes/index.rst).
  - [x] Remove bold format from title.
- [x] Main `index.rst` updates
  - [x] Update link to release note section to use new
         `refid` = `release-notes-collection`.
  - [x] Update _toctree_ specification with path to `release-notes/index`
- [x] Update the [`exclude_patterns` specification](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/conf.py#L100) in the Sphinx configuration file `conf.py` to add the `release-notes/releases` pathr.

After you complete the updates, rebuild the Sphinx project. Resolve any errors. Then, review the local build output to verify that the content renders correctly and that it matches the [API Guide Release notes template](https://github.com/rackerlabs/docs-common/tree/master/api-guide-template/release-notes).
## Root index.rst updates

Many of the refactoring changes in the root `index.rst` file were completed when you refactored earlier section.  These instructions provide a summary of the changes so you can verify them.
- [x] Ensure the first paragraph provides information  about audience and prerequisite knowledge like the [root index.rst template](https://github.com/rackerlabs/docs-common/blob/master/api-guide-template/index.rst).
- [x] Update the links to API Guide sections
  - [x] Remove Developer-Guide
  - [x] Change release-notes ref ID to `release-notes-collection`
- [x] Update files in `toctree` directive to follow this pattern:

```
   Product name v# <https://developer.rackspace.com/docs/cloud-load-balancers/v1/>
   getting-started/index
   general-api-info/index
   api-reference/index
   release-notes/index
   service-updates
   additional-resources
   copyright
```

Changes include: 
- [x] Add the following file names:
  - [x]  `additional-resources.rst`
  - [x] `service-updates.rst`
- [x] Delete the following file names: 
  - [x] getting-started.rst topic
  - [x] prerequisites-for-using-api.rst
  - [x] concepts.rst
  - [x] developer-guide.rst
  - [x] api-reference.rst

**Note:** If your project contains additional source files not included in this list, add them to [Issue #62 in the docs-common repo](https://github.com/rackerlabs/docs-common/issues/62). Include them in the content architecture where it makes sense. IA team can review.
## Next steps

After you complete the refactoring work, complete the following steps to submit the changes for review:
1. To change the API documentation template for your project, update the [routes.d](https://github.com/rackerlabs/nexus-control/blob/master/config/routes.d/developer.rackspace.com.json) file in the nexus-control repository to change the template from `docs-singlepage.html` to `user-guide.html`.
2. After the nexus-control PR has been merged, submit a PR with your refactoring changes against your upstream repository.
3. Follow the [review instructions](https://github.com/rackerlabs/docs-common/blob/master/refactor-api-guide.md#review).

**Comment by cyrichardson** [comment]
Closes #120

**Comment by catlook** [comment]
related PRs: rackerlabs/docs-quickstart#116
rackerlabs/nexus-control#560
